packages/commonwealth/client/scripts/controllers/server/contracts.ts

```diff
@@ -15,12 +21,160 @@ class ContractsController {
+    return this._store.getCommunityContracts();
+  }
+
+  public async addContractAbi(
+    contract: Contract,
+    abi: Array<Record<string, unknown>>
+  ) {
+    const response: TypedResponse<CreateContractAbiResp> = await $.post(
+      `${app.serverUrl()}/createContractAbi`,
+      {
+        jwt: app.user.jwt,
+        contractId: contract.id,
+        abi,
+      }
+    );
+    const resultContract: ContractAttributes = response['result']['contract'];
+    const resultAbi: ContractAbiAttributes = response['result']['contractAbi'];
+    this.update(resultAbi.abi, resultContract);
+    return response;
+  }
+
+  public async checkFetchEtherscanForAbi(address: string) {
+    try {
+      const response: Response = await $.post(
+        `${app.serverUrl()}/etherscanAPI/fetchEtherscanContract`,
+        {
+          address,
+        }
+      );
+      console.log(response);
+      const resultContract: ContractAttributes = response['result']['contract'];
+      const resultAbi: ContractAbiAttributes =
+        response['result']['contractAbi'];
+      this.update(resultAbi.abi, resultContract);
+    } catch (err) {
+      console.log('Failed to fetch abi from etherscan', err);
+      throw new Error(err);
+    }
+  }
+
+  public async update(
+    contractAbi: Array<Record<string, unknown>>,
+    contractAttributes: ContractAttributes
+  ) {
+    // Update contract in store
+    if (this._store.getById(contractAttributes.id)) {
+      this._store.remove(this._store.getById(contractAttributes.id));
+    }
+    const {
+      id,
+      address,
+      chain_node_id,
+      type,
+      createdAt,
+      updatedAt,
+      decimals,
+      token_name,
+      symbol,
+      is_factory,
+      nickname,
+    } = contractAttributes;
+    this._store.add(
+      new Contract(
+        id,
+        address,
+        chain_node_id,
+        type,
+        createdAt,
+        updatedAt,
+        decimals,
+        token_name,
+        symbol,
+        contractAbi,
+        is_factory,
+        nickname
+      )
+    );
+  }
+
+  public async add(
+    community: string,
+    balance_type: BalanceType,
+    chain_node_id: number,
+    node_url: string,
+    address: string,
+    abi: string,
+    contractType: ContractType,
+    symbol: string,
+    token_name: string,
+    decimals: number
+  ) {
+    try {
+      const response: TypedResponse<CreateContractResp> = await $.post(
+        `${app.serverUrl()}/createContract`,
+        {
+          community,
+          balance_type,
+          chain_node_id,
+          jwt: app.user.jwt,
+          node_url,
+          address,
+          abi,
+          contractType,
+          symbol,
+          token_name,
+          decimals,
+        }
+      );
+      const responseContract: ContractAttributes =
+        response['result']['contract'];
+      const { id, type, is_factory, nickname, createdAt, updatedAt } =
+        responseContract;
+      const result = new Contract(
+        id,
```

Comment:
Named arguments here too -- lots of wasted / weird duplicate destructuring code.